src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -84,55 +85,54 @@ internal JsonClassInfo(Type type, JsonSerializerOptions options)
+                                ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameNull(this, jsonPropertyInfo);
+                            }
 
-                        jsonPropertyInfo.ClearUnusedValuesAfterAdd();
-                    }
-                }
-            }
-            else if (ClassType == ClassType.Enumerable || ClassType == ClassType.Dictionary)
-            {
-                // Add a single property that maps to the class type so we can have policies applied.
-                JsonPropertyInfo jsonPropertyInfo = AddPolicyProperty(type, options);
+                            // If the JsonPropertyNameAttribute or naming policy results in collisions, throw an exception.
+                            if (!propertyNames.Add(jsonPropertyInfo.NameUsedToCompareAsString))
+                            {
+                                ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameConflict(this, jsonPropertyInfo);
+                            }
 
-                // Use the type from the property policy to get any late-bound concrete types (from an interface like IDictionary).
-                CreateObject = options.ClassMaterializerStrategy.CreateConstructor(jsonPropertyInfo.RuntimePropertyType);
+                            jsonPropertyInfo.ClearUnusedValuesAfterAdd();
+                        }
+                    }
+                    break;
+                case ClassType.Enumerable:
+                case ClassType.Dictionary:
+                    // Add a single property that maps to the class type so we can have policies applied.
+                    JsonPropertyInfo policyProperty = AddPolicyProperty(type, options);
 
-                // Create a ClassInfo that maps to the element type which is used for (de)serialization and policies.
-                Type elementType = GetElementType(type);
+                    // Use the type from the property policy to get any late-bound concrete types (from an interface like IDictionary).
+                    CreateObject = options.ClassMaterializerStrategy.CreateConstructor(policyProperty.RuntimePropertyType);
 
-                ElementClassInfo = options.GetOrAddClass(elementType);
-            }
-            else if (ClassType == ClassType.Value)
-            {
-                // Add a single property that maps to the class type so we can have policies applied.
-                AddPolicyProperty(type, options);
-            }
-            else
-            {
-                Debug.Assert(ClassType == ClassType.Unknown);
-                // Do nothing. The type is typeof(object).
+                    // Create a ClassInfo that maps to the element type which is used for (de)serialization and policies.
+                    Type elementType = GetElementType(type);
+                    ElementClassInfo = options.GetOrAddClass(elementType);
+                    break;
+                case ClassType.Value:
+                case ClassType.Unknown:
```

Comment:
Outside of making this a case statement, treating Unknown like Value is the only functional change here.